certora/specs/LiquidateBuffer.spec

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+using Util as Util;
+
+methods {
+    function extSloads(bytes32[]) external returns (bytes32[]) => NONDET DELETE;
+
+    function lastUpdate(MorphoHarness.Id) external returns (uint256) envfree;
+    function borrowShares(MorphoHarness.Id, address) external returns (uint256) envfree;
+    function collateral(MorphoHarness.Id, address) external returns (uint256) envfree;
+    function totalBorrowShares(MorphoHarness.Id) external returns (uint256) envfree;
+    function totalBorrowAssets(MorphoHarness.Id) external returns (uint256) envfree;
+    function virtualTotalBorrowAssets(MorphoHarness.Id) external returns uint256 envfree;
+    function virtualTotalBorrowShares(MorphoHarness.Id) external returns uint256 envfree;
+
+    function Util.libId(MorphoHarness.MarketParams) external returns (MorphoHarness.Id) envfree;
+    function Util.lif(uint256) external returns (uint256) envfree;
+    function Util.oraclePriceScale() external returns (uint256) envfree;
+    function Util.wad() external returns (uint256) envfree;
+
+    function Morpho._isHealthy(MorphoHarness.MarketParams memory, MorphoHarness.Id,address) internal returns (bool) => NONDET;
+    function Morpho._accrueInterest(MorphoHarness.MarketParams memory, MorphoHarness.Id) internal => NONDET;
+
+    function _.price() external => constantPrice expect uint256;
+}
+
+persistent ghost uint256 constantPrice;
+
+// Check that for a position with LTV < 1 / LIF, its health improves after a liquidation.
+rule liquidateImprovePosition(env e, MorphoHarness.MarketParams marketParams, address borrower, uint256 seizedAssetsInput, uint256 repaidSharesInput, bytes data) {
+    // Assume no callback.
+    require data.length == 0;
+
+    MorphoHarness.Id id = Util.libId(marketParams);
+
+    // We place ourselves at the last block for getting the following variables.
+    require lastUpdate(id) == e.block.timestamp;
+
+    uint256 borrowerShares = borrowShares(id, borrower);
+    // Safe require because of the sumBorrowSharesCorrect invariant.
+    require borrowerShares <= totalBorrowShares(id);
```

Comment:
could we requireInvariant here?

Reply:
We could not do it directly here, because the invariant is not saying this explicitly. Rather it is showing that totalBorrowShares is equal to the ghost value that is updated each time a borrowShares gets changed. We could use some of the same tricks that we used in morpho-token, with partial sums, to make that explicit. But it would require some significant change to the spec, so I'm not sure if it is worth it at this point